wallet-new/src/Cardano/Wallet/Kernel/Migration.hs

```diff
@@ -0,0 +1,229 @@
+                        <> " rootAddress(es) to migrate."
+
+                mapM_ (restore logMsg pw keystore) available
+
+           -- Now that we have closed the DB, we can move the directory
+           backupPath <- moveLegacyDB legacyDbPath
+
+           -- asynchronous restoration still runs at this point.
+           logMsg Info $ "acid state migration succeeded. Old db backup can be found at " <> pack backupPath
+
+
+restore :: (Severity -> Text -> IO ())
+        -> Kernel.PassiveWallet
+        -> Keystore
+        -> MigrationMetadata
+        -> IO ()
+restore logMsg pw keystore metadata = do
+    let wId = WalletIdHdRnd (metadata ^. mmHdRoot . HD.hdRootId)
+        -- Our best guess whether or not this wallet has the spending password.
+        -- currently we always yield True, but we can tweak these values if
+        -- our assumptions reveal not to be correct (or if we can fine tune
+        -- our heuristic).
+        bestGuess = case metadata ^. mmHasSpendingPasswordHeuristic of
+                         LowConfidence  -> True
+                         HighConfidence -> True
+    mEsk <- Keystore.lookup wId keystore
+    case mEsk of
+        Just esk -> do
+            res <- restoreWallet pw
+                                 bestGuess
+                                 (metadata ^. mmDefaultAddress)
+                                 (metadata ^. mmHdRoot . HD.hdRootName)
+                                 (metadata ^. mmHdRoot . HD.hdRootAssurance)
+                                 esk
+                                 (prefilter esk pw wId)
+            case res of
+                 Right (restoredRoot, balance) -> do
+                     let errMsg = "Migrating " % F.build
+                                % " with balance " % F.build
+                     logMsg Info (F.sformat errMsg restoredRoot balance)
+                 Left err -> do
+                     let errMsg = "Couldn't migrate " % F.build
+                                % " due to : " % F.build
+                     logMsg Error (F.sformat errMsg wId err)
+        Nothing -> do
+            let errMsg = "Couldn't migrate " % F.build
+                       % " : the key was not found in the keystore."
+            logMsg Error (F.sformat errMsg wId)
+
+-- PRECONDITION: The 'FilePath' should exist. This is checked at the call site.
+bracketLegacyDB :: FilePath -> (WS.WalletStorage -> IO a) -> IO a
+bracketLegacyDB legacyDbPath withWalletStorage =
+    bracket (openState False legacyDbPath)
+            closeState
+            -- Apparently hlint thinks this is more readable, heh.
+            (getWalletSnapshot >=> withWalletStorage)
+
+-- | Move the legacy database into a backup directory.
+moveLegacyDB :: FilePath -> IO FilePath
+moveLegacyDB filepath = do
+    now <- getCurrentTime
+    let backupPath =  filepath
+                   <> "-backup-"
+                   <> formatTime defaultTimeLocale (iso8601DateFormat (Just "%H:%M:%S")) now
```

Comment:
Having these special characters in filenames make me a bit worried... maybe try to avoid?